### PR TITLE
XYPanel: Improvements

### PR DESCRIPTION
**Pull request by mdvictor**
- [x] Show `prepScatter` errors to user
- [x] Fix explicit mode to support multiple 'layers' (see geomap) 
- [x] Improve tooltip and apply tooltip options
- [x] Improve legend and apply legend options
- [ ] Weird lines appear on refresh in some cases (gdev dashboard shows this happening)
- [x] rename XY - Explicit to Auto - Manual
- [x] Enable data links
- [x] Migrated PureComponent to FunctionalComponent

More points TBD

### CODE REVIEW

**Comment by codeincarnate** [comment]
Part of #54192

**Comment by mdvictor** [comment]
TODO: rename XY - Explicit to Auto - Manual

**Comment by mdvictor** [comment]
test data links and annotations

**Comment by codeincarnate** [comment]
Nitpicking but it  looks like the label background is off here:

https://user-images.githubusercontent.com/199847/188040862-adcf8839-4635-42ce-9044-e2c2c2370872.mov

**Comment by codeincarnate** [comment]
Looks like series selection with the legend isn't working 😢 

https://user-images.githubusercontent.com/199847/188041552-0cbe6541-3b82-4fa4-8739-9e65214856ec.mov

File: public/app/plugins/panel/xychart/XYChartPanel2.tsx [hunk 1]
```diff
@@ -13,116 +22,225 @@ import {
...
+  const [series, setSeries] = useState<ScatterSeries[]>([]);
+  const [builder, setBuilder] = useState<UPlotConfigBuilder | undefined>();
+  const [facets, setFacets] = useState<FacetedData | undefined>();
+  const [hover, setHover] = useState<ScatterHoverEvent | undefined>();
+  const [shouldDisplayCloseButton, setShouldDisplayCloseButton] = useState<boolean>(false);
 
-  componentDidMount() {
-    this.initSeries(); // also data
-  }
+  const isToolTipOpen = useRef<boolean>(false);
```
Comment (codeincarnate) [inline public/app/plugins/panel/xychart/XYChartPanel2.tsx:43]: This needs to be a ref?

File: public/app/plugins/panel/xychart/XYChartPanel2.tsx [hunk 2]
```diff
@@ -13,116 +22,225 @@ import {
...
+      </div>
     );
   }
-}
+
+  return (
+    <>
+      <VizLayout width={props.width} height={props.height} legend={renderLegend()}>
+        {(vizWidth: number, vizHeight: number) => (
+          // <pre style={{ width: vizWidth, height: vizHeight, border: '1px solid green', margin: '0px' }}>
```
Comment (codeincarnate) [inline public/app/plugins/panel/xychart/XYChartPanel2.tsx:199]: I think this can be killed 😄

File: public/app/plugins/panel/xychart/config.ts [hunk 3]
```diff
@@ -34,16 +32,16 @@ export function getScatterFieldConfig(cfg: ScatterFieldConfig): SetFieldConfigOp
         .addRadio({
           path: 'point',
           name: 'Points',
-          category: categoryStyles,
+          category: undefined,
```
Comment (codeincarnate) [inline public/app/plugins/panel/xychart/config.ts:35]: Should this be removed as opposed to undefined?

**Review by codeincarnate — APPROVED** [review]
Epic PR, lots of really great changes. I think we should keep in mind my comments (of course 😆) but I don't see any reason this shouldn't be merged currently.

File: public/app/plugins/panel/xychart/XYChartPanel2.tsx (hunk 1 above)
Comment (mdvictor) [reply public/app/plugins/panel/xychart/XYChartPanel2.tsx:43]: I tried with state but it didn't work out. Not entirely sure why but I'm going to leave it a ref for now and look over it later.

File: public/app/plugins/panel/xychart/ManualEditor.tsx [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+  if (options === undefined || !options.series) {
+    return null;
+  }
+
+  return (
+    <>
+      <Button icon="plus" size="sm" variant="secondary" onClick={createNewSeries}>
+        Add series
+      </Button>
+      <br />
```
[the remark this answers is not in the document]
Comment (mdvictor) [reply public/app/plugins/panel/xychart/ManualEditor.tsx:105]: Modified the html

File: public/app/plugins/panel/xychart/ManualEditor.tsx [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+export const ManualEditor: FC<StandardEditorProps<ScatterSeriesConfig[], any, XYChartOptions>> = ({
+  value,
+  onChange,
+  context,
+}) => {
+  const [selected, setSelected] = useState<number>(-1);
+  const style = useStyles(getStyles);
+
+  const onXFieldChange = (x: string | undefined, index: number) => {
```
[the remark this answers is not in the document]
Comment (mdvictor) [reply public/app/plugins/panel/xychart/ManualEditor.tsx:21]: Indeed. Fixed.

**Comment by mdvictor** [comment]
> Looks like series selection with the legend isn't working 😢
> 
>  Screen.Recording.2022-09-01.at.8.35.54.PM.mov

Yep, that was the last thing to be done. Pushing it asap.

**Comment by mdvictor** [comment]
> Nitpicking but it looks like the label background is off here:
> 
>  Screen.Recording.2022-09-01.at.8.29.38.PM.mov

fixed

**Comment by mdvictor** [comment]
Fixed legend default position to be on left (like in other panels)
Introduced legend toggle on label click.
Fixed (most) CR comments

There are still some issues to be fixed that I have not yet pinpointed exactly.
Changing datasources sometimes causes errors to be thrown that should be caught and handled.
CSV Content mode in the TestDatasource crashes the browser entirely (something is very wrong with the panel here)

I would like to fix these issues before merging

**Pull request merged**